File: platforms/macos/src/Core/PromptBuilder.swift
import Foundation

/// Substitutes the source text into the configurable prompt template (spec §5).
///
/// Behaviour (pinned by `conformance/prompt-builder.json`):
/// - empty template ⇒ return content verbatim
/// - template contains `{content}` ⇒ replace the placeholder
/// - no placeholder ⇒ append the content after the rules, separated by `\n\n`
public enum PromptBuilder {
    public static let placeholder = "{content}"
    public static let targetPlaceholder = "{target}"

    public static func build(template: String, content: String) -> String {
        if template.isEmpty {
            return content
        }
        if let range = template.range(of: placeholder) {
            return template.replacingCharacters(in: range, with: content)
        }
        return template + "\n\n" + content
    }

    /// Resolve the unified target language into the template's `{target}` placeholder. Done ONCE (before
    /// `build` substitutes `{content}`), so every prompt-driven backend shares the same target. A template
    /// without `{target}` is returned unchanged. Mirrors the Windows `PromptBuilder.WithTarget`.
    public static func withTarget(_ template: String, _ target: String?) -> String {
        template.isEmpty ? template : template.replacingOccurrences(of: targetPlaceholder, with: target ?? "")
    }
}

File: platforms/windows/src/TranslateTheDamn.Core/Util/PromptBuilder.cs
namespace TranslateTheDamn.Core.Util;

/// <summary>Substitutes the source text into the configurable prompt template.</summary>
public static class PromptBuilder
{
    public const string Placeholder = "{content}";
    public const string TargetPlaceholder = "{target}";

    /// <summary>Resolve the unified target language into the template's <c>{target}</c> placeholder.
    /// Done ONCE at registry build time (not per request). Templates without <c>{target}</c> are unchanged.</summary>
    public static string WithTarget(string template, string? target) =>
        string.IsNullOrEmpty(template) ? template : template.Replace(TargetPlaceholder, target ?? string.Empty);

    public static string Build(string template, string content)
    {
        if (string.IsNullOrEmpty(template)) return content;
        if (template.Contains(Placeholder, StringComparison.Ordinal))
            return template.Replace(Placeholder, content);
        // No placeholder present: append the content after the rules.
        return template.TrimEnd() + "\n\n" + content;
    }
}
